src/ring_theory/graded_algebra/basic.lean

```diff
@@ -103,6 +103,42 @@ dfinsupp.mem_support_iff.trans add_submonoid_class.coe_eq_zero.not.symm
 
 end graded_ring
 
+section add_left_cancel_monoid
+
+open direct_sum dfinsupp finset function
+
+lemma direct_sum.decompose_mul_add_of_left_mem {ι σ A}
+  [decidable_eq ι] [add_left_cancel_monoid ι] [semiring A]
+  [set_like σ A] [add_submonoid_class σ A] (𝒜 : ι → σ) [graded_ring 𝒜]
+  {i j : ι} (a : 𝒜 i) {b : A} :
+  decompose 𝒜 (↑a * b) (i + j) =
+    @graded_monoid.ghas_mul.mul ι (λ i, 𝒜 i) _ _ _ _ a (decompose 𝒜 b j) :=
+subtype.ext $ or.elim (em (a = 0)) (λ EQ, EQ.symm ▸ by simp) $ λ INEQ, begin
+  classical,
+  simp_rw [decompose_mul, coe_mul_apply, decompose_coe, support_of _ i a INEQ, singleton_product,
+    map_filter, sum_map, comp, embedding.coe_fn_mk, add_left_cancel_iff, filter_eq'],
+  refine dite (decompose 𝒜 b j = 0) (λ h, by simp [if_neg (not_mem_support_iff.mpr h), h]) (λ h, _),
+  erw [if_pos (mem_support_iff.mpr h), finset.sum_singleton, of_eq_same],
+  refl,
+end
+
+lemma direct_sum.decompose_mul_add_of_right_mem {ι σ A}
+  [decidable_eq ι] [add_right_cancel_monoid ι] [semiring A]
```

Comment:
This probably shouldn't be in the `add_left_cancel_monoid` section!

Reply:
indeed, but instead of creating a new section, I just renamed the section as add_cancel_monoid